### PR TITLE
enhance cache expiry

### LINKED ISSUE #95: Ignite cache cannot expire
**Questions**
Set the cache and set the expiration time. When the expiration time is exceeded, the cache expires.
However, set the cache and set the expiration time. If you get the cache, the cache will not expire. This is a very serious problem. I found that the expiration policy was cleared when getting the cache.
image

**Version**
vertx-ignite 3.9.0 or laster.

### PR DESCRIPTION
**Pull request by zyclonite**
- remove deprecated defaultLockTimeout from cache options
- add expiryPolicy to cache options
- do not override expiryPolicy for cache get when it is already set by configuration, fixes #95

### CODE REVIEW

**Comment by zyclonite** [comment]
@tsegismont do you think it would make sense to add this test to the suite? [testMapPutThenPutTtl](https://github.com/zyclonite/vertx-ignite/blob/enhance_cache_expiration/src/test/java/io/vertx/core/shareddata/IgniteClusteredAsyncMapTest.java#L50-L62)

**Comment by zyclonite** [comment]
this is the vert.x core pr https://github.com/eclipse-vertx/vert.x/pull/3730

File: src/main/java/io/vertx/spi/cluster/ignite/impl/SubsMapHelper.java [hunk 1]
```diff
@@ -88,10 +88,10 @@ public void remove(String address, RegistrationInfo registrationInfo, Promise<Vo
   }
 
   public void removeAllForNode(String nodeId) {
-    Set<IgniteRegistrationInfo> toRemove = map.query(new ScanQuery<IgniteRegistrationInfo, Boolean>((k, v) -> k.registrationInfo().nodeId().equals(nodeId)))
+    TreeSet<IgniteRegistrationInfo> toRemove = map.query(new ScanQuery<IgniteRegistrationInfo, Boolean>((k, v) -> k.registrationInfo().nodeId().equals(nodeId)))
```
Comment (tsegismont) [inline src/main/java/io/vertx/spi/cluster/ignite/impl/SubsMapHelper.java:91]: Is this related to the issue?

File: src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java [hunk 2]
```diff
@@ -34,6 +35,14 @@ protected ClusterManager getClusterManager() {
     return new IgniteClusterManager();
   }
 
+  @Override
+  protected List<String> getExternalNodeSystemProperties() {
+    return Arrays.asList(
+      "-Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.SLF4JLogDelegateFactory",
```
Comment (tsegismont) [inline src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java:41]: This one is not needed anymore in Vert.x 4

File: src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java [hunk 3]
```diff
@@ -36,11 +36,6 @@ public static void fromJson(Iterable<java.util.Map.Entry<String, Object>> json,
             obj.setCopyOnRead((Boolean)member.getValue());
           }
           break;
-        case "defaultLockTimeout":
```
Comment (tsegismont) [inline src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java:39]: If you remove this, don't forget to update the breaking changes wiki page.

**Review by tsegismont — COMMENTED** [review]
@zyclonite LGTM, thanks! Just a few comments/concerns

File: src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java (hunk 3 above)
Comment (zyclonite) [reply src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java:39]: i removed this because it was deprecated already and had no effect

File: src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java (hunk 2 above)
Comment (zyclonite) [reply src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java:41]: copied this from here https://github.com/vert-x3/vertx-hazelcast/blob/master/src/test/java/io/vertx/core/eventbus/HazelcastFaultToleranceTest.java#L50

but it did improve the logging behaviour, it seemed to be required only for this test suite
did i miss something?

File: src/main/java/io/vertx/spi/cluster/ignite/impl/SubsMapHelper.java (hunk 1 above)
Comment (zyclonite) [reply src/main/java/io/vertx/spi/cluster/ignite/impl/SubsMapHelper.java:91]: not related but could/does lead to deadlocks and makes the tests unstable... shall i move it into a separate pr?

File: src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java (hunk 3 above)
Comment (tsegismont) [reply src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java:39]: Fine with me. Don't forget to add it to https://github.com/vert-x3/wiki/wiki/4.0.1-Deprecations-and-breaking-changes

File: src/main/java/io/vertx/spi/cluster/ignite/impl/SubsMapHelper.java (hunk 1 above)
Comment (tsegismont) [reply src/main/java/io/vertx/spi/cluster/ignite/impl/SubsMapHelper.java:91]: Ok, thanks for the details. Ideally yes, if you can, prefer splitting separate concerns. But no biggie if you keep it here.

File: src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java (hunk 2 above)
Comment (tsegismont) [reply src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java:41]: Interesting, can you describe what this sysprop did exactly?

In Vert.x 4 it should no longer be needed, there is now a detection mechanism which picks a log delegate factory corresponding to the logging implementation available in the classpath.

Comment (tsegismont) [reply src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java:41]: Closing this thread because it's no reason to block the PR but let's dig into this further. I'd like to make sure we're not missing something.

Comment (tsegismont) [reply src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java:41]: @zyclonite I had a closer look and it seems Vert.x Web test jar has a `vertx-default-jul-logging.properties` file. When this file is present in the classpath, it forces Vert.x to use jdk logging, unless the sysprop is set.

File: src/main/java/io/vertx/spi/cluster/ignite/impl/SubsMapHelper.java (hunk 1 above)
Comment (zyclonite) [reply src/main/java/io/vertx/spi/cluster/ignite/impl/SubsMapHelper.java]: did that now, no problem as long it's not touching the same files ;)

File: src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java (hunk 3 above)
Comment (zyclonite) [reply src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java:39]: ok, will do

File: src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java (hunk 2 above)
Comment (zyclonite) [reply src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java:41]: ok, so i'll take a note to fix that when vertx-web is fixed

Comment (tsegismont) [reply src/test/java/io/vertx/core/eventbus/IgniteFaultToleranceTest.java:41]: Actually we'll have to keep this because Vert.x Web will keep using the `vertx-default-jul-logging.properties`. But no worries, I just wanted to make sure there was no bug with logger factory selection, and there isn't.

File: src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java (hunk 3 above)
Comment (tsegismont) [reply src/main/generated/io/vertx/spi/cluster/ignite/IgniteCacheOptionsConverter.java:39]: We have completed the vote for adding you as committer. Soon you will be able to create/edit in the wiki

**Review by tsegismont — APPROVED** [review]
LGTM, thanks!
